### PR TITLE
feat: staged diagram generation in automator

### PR DESCRIPTION
**Pull request by heleaf**
Screen Shot 2021-06-28 at 12 46 56 PM
Screen Shot 2021-06-28 at 12 54 53 PM

# Description

Related issue/PR: #42 

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Change Summary: 
- Added a new command to `automator` that allows for command-line SVG generation of a single sub/sty/dsl trio.
- Added an option to `automator` that allows for command-line generation of staged diagrams (as multiple SVG files) of sub/sty/dsl trios. 

Motivation:
- `automator` was missing functionality to generate SVG(s) for a single sub/sty/dsl trio (previously mostly working on batch sub/sty/dsl) 
- Staged diagrams have been requested alongside the final diagram itself by users 

Dependencies: nothing new

# Implementation strategy and design decisions

Include a high-level summary of the implementation strategy and list important design decisions made, if any.
- Changes to `automator` can be found in `index.tsx`.
- The functions used to generate states for staged diagrams can be found in `Comic.ts` under `core/src/renderer`.

### Staged diagram generation functions
- The staged diagram states are determined as a post-processing calculation (i.e. after the state of the original sub/sty/dsl has been fully optimized). Each state has certain missing shapes from their shapelist according to the order in which the user declared GPI primitives in their substance program. 
- The main function for creating the list of states is `getListOfStagedStates` in `Comic.ts`. 
- `getListOfStagedStates` creates a list of objects with FGPI properties from a state's `translation.trMap` and uses that list to determine what GPIs should remain in a state's shapelist for a list of staged states. 
- The order in which shapes get "added" state-by-state depends on the order in which the user defines GPI-tagged objects in substance. Note: whether or not substance commands like `Set A` and `IsSubset(B,A)` have GPI properties or not depends on the definitions given in style. For example, if the user's substance program is `Set A, B, C, IsSubset(B,A)` and their style program associates each `Set` with a `Text` label, and `IsSubset` with an `Arrow`,  then the shapelists in the list of generated states would look something like `[[Text for Set A], [Text for Set A, Text for Set B], [Text for Set A, Text for Set B, Text for Set C], [Text for Set A, Text for Set B, Text for Set C, Arrow for IsSubset(B,A)]]`.
Screen Shot 2021-06-28 at 12 47 34 PM

- When called in `automator`'s `index.tsx`, this function is used to generate multiple SVG files, corresponding to each state in the outputted list of states. 

### `automator` changes
- A new command line call (`draw`) has been added that allows the user to generate a single SVG file that corresponds to a single sub/sty/dsl trio. See line 29 of `const USAGE` for details. The call to `draw` is located in the `else` case of the `async ()` call on line 323. 
- A new command line option (`--staged`) has been added that allows the user to generate staged SVG files that correspond to a single sub/sty/dsl trio. See line 36 of `const USAGE` for details. 
Screen Shot 2021-06-28 at 12 46 16 PM

- Function `singleProcess` has a new parameter `staged` and several new cases to account for the `--staged` option. `batchProcess` also has a new parameter `staged` for this option. 

# Examples with steps to reproduce them
- Run `automator draw tree.sub tree.sty setTheory.dsl out --src-prefix=../../examples/set-theory-domain` in `automator` to generate an example SVG for a sub/sty/dsl trio. 
- Add a `--staged` flag to the above command (i.e. run automator draw tree.sub tree.sty setTheory.dsl out --src-prefix=../../examples/set-theory-domain --staged`) to generate staged SVGs for a sub/sty/dsl trio. 

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass locally using `yarn test`
- [x] I ran `yarn docs` and there were no errors when generating the HTML site
- [x] My code follows the style guidelines of this project (e.g.: no ESLint warnings)

# Open questions
Questions that require more discussion or to be addressed in future development:  
- With the $LOCAL variables that are created for each matched substance statement + style block, it's difficult to tell which $LOCAL variables correspond to each substance statement (which is not great for controlled storyboarding)
- The generated svg window size is weirdly vertical (how can this be controlled?) (this may also be a result of `State` not having a `canvas` property in `automator`). 
- For comic-like staging, it might be possible to treat each generated SVG as a rectangle, and then do some optimization on how to lay that out.

### CODE REVIEW

File: packages/automator/README.md [hunk 1]
```diff
@@ -10,12 +10,14 @@ Penrose Automator.
 Usage:
   automator batch LIB OUTFOLDER [--folders]  [--src-prefix=PREFIX] [--repeat=TIMES] [--render=OUTFOLDER]
   automator render ARTIFACTSFOLDER OUTFOLDER
+  automator draw SUBSTANCE STYLE DOMAIN OUTFOLDER [--folders] [--src-prefix=PREFIX] [--staged]
```
Comment (wodeni) [inline packages/automator/README.md:13]: Not sure if we want the `--folder` option here. Consider removing it

File: packages/automator/README.md [hunk 2]
```diff
@@ -39,3 +41,8 @@ In addition to batch-processing Penrose programs, you can also use `automator` t
 - Open `browser/index.html` to view the result.
 
 ![](docs/penrose-artifacts.png)
+
+## 06/25/2021 Update: Staged Diagram Generation
```
Comment (wodeni) [inline packages/automator/README.md:45]: Don't have to date this :P. A quick screenshot would be great!

File: packages/automator/index.tsx [hunk 3]
```diff
@@ -8,6 +8,8 @@ import {
   stepUntilConvergence,
 } from "@penrose/core";
 import { renderArtifacts } from "./artifacts";
+import { getListOfStagedStates } from "../core/src/renderer/Staging";
```
Comment (wodeni) [inline packages/automator/index.tsx:11]: Kinda surprised that this worked. Export your functions in `index.tsx` of `core` and import it here from `@penrose/core`. These are separate packages, so relative paths won't work in general.

File: packages/automator/index.tsx [hunk 4]
```diff
@@ -175,8 +207,21 @@ const singleProcess = async (
     // returning metadata for aggregation
     return { metadata, state: optimizedState };
   } else {
-    fs.writeFileSync(out, canvas);
-    console.log(chalk.green(`The diagram has been saved as ${out}`));
+    if (staged) {
+      // write multiple svg files out
+      const writeFileOut = (canvasData: any, index: number) => {
+        let filename = out.slice(0, out.indexOf("svg") - 1);
```
Comment (wodeni) [inline packages/automator/index.tsx:213]: Perhaps use [`path.join`](https://nodejs.org/api/path.html#path_path_join_paths) and templated strings for readability

File: packages/automator/index.tsx [hunk 5]
```diff
@@ -222,13 +268,14 @@ const batchProcess = async (
     // try to render the diagram
     try {
       // Warning: will face id conflicts if parallelism used
-      const res = await singleProcess(
+      let res = await singleProcess(
```
Comment (wodeni) [inline packages/automator/index.tsx:271]: Why the `let` here?

File: packages/automator/index.tsx [hunk 6]
```diff
@@ -222,13 +268,14 @@ const batchProcess = async (
...
         subURI,
         styURI,
         dslURI,
         folders,
         `${out}/${name}-${id}${folders ? "" : ".svg"}`,
         prefix,
+        staged, // make something else to fix singleProcess?
```
Comment (wodeni) [inline packages/automator/index.tsx:278]: Is there something wrong with `singleProcess`?

File: packages/automator/index.tsx [hunk 7]
```diff
@@ -273,27 +320,31 @@ const batchProcess = async (
...
-      await batchProcess(args.LIB, folders, args.OUTFOLDER, prefix);
+      await batchProcess(args.LIB, folders, args.OUTFOLDER, prefix, staged);
     }
     if (browserFolder) {
       renderArtifacts(args.OUTFOLDER, browserFolder);
     }
   } else if (args.render) {
     renderArtifacts(args.ARTIFACTSFOLDER, args.OUTFOLDER);
   } else {
+    // this assumes that yarn start draw was called.
```
Comment (wodeni) [inline packages/automator/index.tsx:339]: Now that I think about this, I think it's probably better if we do `else if (args.draw)` and return an error on `else`.

File: packages/core/src/renderer/Staging.ts [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+      throw new Error("shape property doesn't have a .");
+    }
+    const shapeName = shapePropPathName.slice(0, dotIndex);
+    return shapeNamesToInclude.includes(shapeName);
+  };
+
+  const newShapeList = state.shapes.filter(includeShape);
+
+  // to be cleaner, i should technically update shapeOrdering as well
+  // but not modifying that doesn't seem to affect the creation of the SVG
```
Comment (wodeni) [inline packages/core/src/renderer/Staging.ts:80]: Something like:
```ts
return {
  ...state,
  shapes: newShapeList
}
```

**Review by wodeni — CHANGES_REQUESTED** [review]
Looking great! Only some minor comments below.

File: packages/automator/README.md (hunk 2 above)
Comment (heleaf) [reply packages/automator/README.md:45]: What do you mean by a screenshot? 😅

File: packages/automator/index.tsx (hunk 6 above)
Comment (heleaf) [reply packages/automator/index.tsx:278]: woops I'm not sure why I made a comment there 😅 `singleprocess` is fine (i think that was when I was doing some testing)

File: packages/automator/index.tsx (hunk 5 above)
Comment (heleaf) [reply packages/automator/index.tsx:271]: (changed it back to `const` 😅 )

**Review by wodeni — APPROVED** [review]
Thanks for addressing the comments! Good to merge now 🥳

**Pull request merged**